### PR TITLE
Add more UT to UfsUrl

### PR DESCRIPTION
**Pull request by YichuanSun**
### What changes are proposed in this pull request?

Add more unit tests of UfsUrlTest. It is part of a huge [PR](https://github.com/Alluxio/alluxio/pull/18008). This PR only contains the UfsUrlTest enrichment.

### Why are the changes needed?

Improve code quality.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 1]
```diff
@@ -82,25 +99,26 @@ public void UfsUrlCtor() {
...
+        ufsUrl.getParentURL().get().getParentURL().get().toString());
     assertEquals("/xy z/a b c", ufsUrl.getFullPath());
     assertEquals("hdfs", ufsUrl.getScheme());
     assertEquals("hdfs://localhost:8020/xy z/a b c/d", ufsUrl.join("/d").toString());
     assertEquals("hdfs://localhost:8020/xy z/a b c", ufsUrl.toString());
   }
 
   @Test
-  public void hashCodeTest() {
+  public void hashCodeTestOfUfsUrl() {
```
Comment (YichuanSun) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:121]:
```suggestion
  public void hashCode() {
```

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 2]
```diff
@@ -118,13 +136,344 @@ public void hashCodeTest() {
...
+
+  /**
+   * Tests the {@link UfsUrl#toString()} method to normalize paths.
+   */
+  @Test
+  public void normalize() {
+    assertEquals("xyz:///", UfsUrl.createInstance("xyz://").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo/").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo/").toString());
+    assertEquals("xyz:///foo", UfsUrl.createInstance("xyz:/foo/").toString());
```
Comment (YichuanSun) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:439]: what are you doing?

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java (hunk 1 above)
Comment (YichuanSun) [reply dora/core/common/src/test/java/alluxio/UfsUrlTest.java:121]: "hashCode" conflicts with the java.lang.Object.
image

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 3]
```diff
@@ -35,21 +42,23 @@ public class UfsUrlTest {
...
   @Test
   public void basicUfsUrl() {
     UfsUrl ufsUrl = UfsUrl.createInstance("abc://localhost:19998/xy z/a b c");
 
-    assertTrue(ufsUrl.getAuthority().isPresent());
-    assertEquals("localhost:19998", ufsUrl.getAuthority().get().toString());
-    SingleMasterAuthority authority = (SingleMasterAuthority) ufsUrl.getAuthority().get();
+    assertNotNull(ufsUrl.getAuthority());
+    assertEquals("localhost:19998", ufsUrl.getAuthority().toString());
+    SingleMasterAuthority authority = (SingleMasterAuthority) ufsUrl.getAuthority();
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:52]: The existing Authority class implementations are more describing the formats of authority in "an alluxio url", like EmbeddedLogicalAuthority and ZookeeperLogicalAuthority etc.

As we are writing a UfsUrl, the authority names should no longer be in those formats. We should be looking at the authority formats for a UFS, like HDFS vs S3(maybe empty). We may need to throw away the old implementations like `EmbeddedLogicalAuthority` etc.

No need to change anything in this PR. Let's migrate one step at a time.

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 4]
```diff
@@ -62,17 +71,25 @@ public void UfsUrlCtor() {
...
     String authority2 = "";
     String path2 = "/testDir/testFile";
     UfsUrl ufsUrl2 = new UfsUrl(scheme2, authority2, path2);
     assertEquals(scheme2, ufsUrl2.getScheme());
-    assertEquals(authority2, ufsUrl2.getAuthority().get().toString());
+    assertEquals(authority2, ufsUrl2.getAuthority().toString());
     assertEquals(path2, ufsUrl2.getFullPath());
     assertEquals(scheme2 + "://" + authority2 + path2, ufsUrl2.toString());
+
+    String ufsUrlString3 = "xyz:testFolder";
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:86]: interesting, i didn't know this format works

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java (hunk 1 above)
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:121]:
```suggestion
  public void hashCodeIdentity() {
```

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 5]
```diff
@@ -118,13 +136,342 @@ public void hashCodeTest() {
...
+      UfsUrl ufsUrl = UfsUrl.createInstance(path);
+      assertEquals(path, ufsUrl.toString());
+    }
+  }
+
+  /**
+   * Tests the {@link UfsUrl#createInstance(String)} constructor for URL
+   * without scheme or with invalid scheme.
+   */
+  // TODO(Yichuan): Bring it back while the improved ufsurl class is merged.
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:162]: agreed, we can add it back later with a better definition of UfsUrl in Alluxio

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 6]
```diff
@@ -118,13 +136,342 @@ public void hashCodeTest() {
...
+
+    UfsUrl uri2 = new UfsUrl(scheme, authority, path);
+    assertEquals(scheme + "://" + authority + path, uri2.toString());
+  }
+
+  /**
+   * Tests the {@link UfsUrl#equals(Object)} method.
+   */
+  @Test
+  public void equals() {
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:206]: Like it doesn't let you write a `hashCode()` I figure this one also overrides `Object.equals()`?
```suggestion
  public void equality() {
```

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 7]
```diff
@@ -118,13 +136,342 @@ public void hashCodeTest() {
...
+
+    assertEquals("foo://bar boo:8080/abc/c",
+        UfsUrl.createInstance("foo://bar boo:8080/abc///c").toString());
+  }
+
+  /**
+   * Tests the {@link UfsUrl#UfsUrl(String, String, String)} constructor to throw an
+   * exception in case an empty path was provided.
+   */
+  @Test(expected = NullPointerException.class)
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:455]: use `assertThrows()` instead?

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 8]
```diff
@@ -118,13 +136,342 @@ public void hashCodeTest() {
...
+  public void constructFromEmptyPathTest2() {
+    new UfsUrl(null, null, null);
+  }
+
+  /**
+   * Tests the {@link UfsUrl#UfsUrl(String, String, String)} constructor to throw an
+   * exception in case an empty path was provided.
+   */
+  @Test(expected = NullPointerException.class)
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:464]: same

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 9]
```diff
@@ -118,13 +136,342 @@ public void hashCodeTest() {
...
+  public void constructFromEmptyPathTest3() {
+    new UfsUrl("file", null, "");
+  }
+
+  /**
+   * Tests the {@link UfsUrl#createInstance(String)} constructor to throw an exception in case an
+   * invalid URI was provided.
+   */
+  @Test(expected = IllegalArgumentException.class)
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:473]: same

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Added a few more comments. Mostly LGTM and I feel it's good to merge after all those comments are resolved. Good work!

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 10]
```diff
@@ -223,11 +223,8 @@ public String getScheme() {
   /**
    * @return the authority of the {@link UfsUrl}
    */
-  public Optional<Authority> getAuthority() {
-    if (!mProto.hasAuthority()) {
-      return Optional.empty();
-    }
-    return Optional.of(Authority.fromString(mProto.getAuthority()));
+  public Authority getAuthority() {
+    return Authority.fromString(mProto.getAuthority());
```
[the remark this answers is not in the document]
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/uri/UfsUrl.java:227]: There is a `NoAuthority` impl for `Authority` that is kinda like `Optional.empty()`. Personally, I think `NoAuthority` is unnecessary and we should replace it with `Optional`.

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 11]
```diff
@@ -223,11 +224,8 @@ public String getScheme() {
   /**
    * @return the authority of the {@link UfsUrl}
    */
-  public Optional<Authority> getAuthority() {
-    if (!mProto.hasAuthority()) {
-      return Optional.empty();
-    }
-    return Optional.of(Authority.fromString(mProto.getAuthority()));
+  public Authority getAuthority() {
```
Comment (dbw9580) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java:227]: Why this change? IMO `Optional<Authority>` is more explicit than `NoAuthority`.

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java [hunk 12]
```diff
@@ -363,7 +357,9 @@ public int getDepth() {
    */
   public String getName() {
     List<String> pathComponents = getPathComponents();
-    Preconditions.checkArgument(!pathComponents.isEmpty());
+    if (pathComponents.isEmpty()) {
+      return "";
```
Comment (dbw9580) [inline dora/core/common/src/main/java/alluxio/uri/UfsUrl.java:361]: why is path components list allowed to be empty? example?

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java (hunk 12 above)
Comment (YichuanSun) [reply dora/core/common/src/main/java/alluxio/uri/UfsUrl.java:361]: It is used to handle url "/". The internal class `Parser` splits raw String to pathComponents by '/', so path "/" is corresponding to an empty pathComponents. Since it is an UfsUrl, may the path like "/" be disappeared? Is it possible that UfsUrl is  local file system with path "/"?

Comment (YichuanSun) [reply dora/core/common/src/main/java/alluxio/uri/UfsUrl.java:361]: It is used to handle url "/". The internal class Parser splits raw String to pathComponents by '/', so path "/" is corresponding to an empty pathComponents. Since it is an UfsUrl, may the path like "/" be disappeared? Is it possible that UfsUrl represents a local file system with path "/"?

File: dora/core/common/src/main/java/alluxio/uri/UfsUrl.java (hunk 11 above)
Comment (YichuanSun) [reply dora/core/common/src/main/java/alluxio/uri/UfsUrl.java]: Have replaced it with `Optional<Authority>`.

**Review by dbw9580 — APPROVED** [review]
LGTM

File: dora/core/common/src/test/java/alluxio/UfsUrlTest.java [hunk 13]
```diff
@@ -118,13 +136,352 @@ public void hashCodeTest() {
...
+  /**
+   * Tests the {@link UfsUrl#UfsUrl(String, String, String)} constructor to throw an
+   * exception in case an empty path was provided.
+   */
+  @Test
+  public void constructFromEmptyPathTest3() {
+    assertThrows(NullPointerException.class, () -> {
+      new UfsUrl("file", null, "");
+    });
+  }
```
Comment (jiacheliu3) [inline dora/core/common/src/test/java/alluxio/UfsUrlTest.java:475]: You may:
1. Either, collapse all these 3 `constructFromEmptyPathTest` cases into one `@Test`
2. Or, rename these 3 like, `nullPath`, `nullAuthority`, `nullScheme` etc

**Review by jiacheliu3 — APPROVED** [review]
LGTM with some minor comments, thanks a lot for the work! Good job :)
